features/70-sdk-api-grouping/api.md

```diff
@@ -0,0 +1,257 @@
+[container.auth forgotPasswordWithEmail:@"email@example.com" completion:nil];
+[container.pubsub subscribeTo:"channel" handler:nil];
+
+SKYRecord *sampleItem = [SKYRecord recordWithRecordType:@"SampleItem"];
+[container.db.public saveRecord:sampleItem completion:nil];
+```
+
+```obj-c
+// Retreieve contexts
+#import <SKYKit/SKYKit.h>
+
+SKYContainer *container = [SKYContainer defaultContainer];
+NSString *apiKey = [container apiKey];
+NSString *currentAccessToken = [container.auth currentAccessToken];
+NSDictionary *publicDBCache = [container.db.public cache];
+```
+
+## Java
+
+```java
+// Call API
+import android.content.Context;
+import io.skygear.skygear.Container;
+import io.skygear.skygear.db.Record;
+
+Container container = Container.defaultContainer(Context.getApplicationContext());
+container.auth().loginWithEmail("email@example.com", "secret");
+container.auth().forgotPassword("email@example.com");
+container.pubsub().subscribe("channel", new Pubsub.Handler());
+
+Record sampleItem = new Record('SampleItem');
+container.db().public().save(sampleItem, null);
+```
+
+```java
+// Retreieve contexts
+import android.content.Context;
+import io.skygear.skygear.Container;
+
+Container container = Container.defaultContainer(Context.getApplicationContext());
+String apiKey = container.apiKey();
+String currentAccessToken = container.auth().currentAccessToken();
+Map publicDBCache = container.db().public().cache();
+```
+
+# Changes on SDK
+
+Proposal:
+
+```
+Package: /
+
+  Container:
+
+    state:
+      url
+      apiKey
+      auth(AuthContainer)
+      user(UserContainer)
+      db(DatabaseContainer)
+      device(DeviceContainer)
+      pubsub(PubsubContainer)
+
+    api:
+      config
+      configApiKey
+      lambda
+
+  Error:
+
+  _Private(can be part of Container):
+
+    state:
+      timeoutOptions
+
+    api:
+      makeUploadAssetRequest
+      sendRequestObject
+      makeRequest
+
+
+Package: /auth
+
+  AuthContainer:
+
+    state:
+      accessToken
+
+    api:
+      signupWithUsername
+      signupWithEmail
+      signupWithUsernameAndProfile
+      signupWithEmailAndProfile
+      signupAnonymously
+      loginWithUsername
+      loginWithEmail
+      loginWithProvider
+      logout
+      changePassword
+      forgotPassword
+
+
+Package: /user
+
+  UserContainer:
+
+    state:
+      user
+      relation(RelationContainer)
+
+    api:
+      saveUser
+      whoami
+      getUsersByEmail
+      getUsersByUsername
+      discoverUserByEmails
+      discoverUserByUsernames
+
+  User:
+
+  RelationContainer:
+
+    api:
+      queryFriend
+      queryFollower
+      queryFollowing
+      add
+      remove
+
+
+Package: /db
+
+  DatabaseContainer:
+
+    state:
+      publicDB(Database)
+      privateDB(Database)
+      cacheResponse(Boolean)
+
+    api:
+      setAdminRole
+      setDefaultRole
+      getDefaultACL
+      setDefaultACL
+      setRecordCreateAccess
+      setRecordDefaultAccess
```

Comment:
@chpapa mentioned that ACL configuration should goes to public db.

Pros: Explicit that public DB support ACL while private DB does not.
Cons: We will have a `PublicDatabase` and `PrivateDatabase` class (instead of just one `Database` class).